llvm/lib/IR/ModuleSummaryIndex.cpp

```diff
@@ -213,9 +211,10 @@ propagateAttributesToRefs(GlobalValueSummary *S,
   for (auto &VI : S->refs()) {
     assert(VI.getAccessSpecifier() == 0 || isa<FunctionSummary>(S));
     if (!VI.getAccessSpecifier()) {
-      if (!MarkedNonReadWriteOnly.insert(VI).second)
+      if (VI.getRef()->second.MarkedNonReadWriteOnly)
         continue;
-    } else if (MarkedNonReadWriteOnly.contains(VI))
```

Comment:
I think this check might be useless and possibly can be removed. We only get here and do the else if check if getAccessSpecifier() is non-zero, which means the VI is still either read or write only. But we only add a VI to the MarkedNonReadWriteOnly set if its getAccessSpecifier() is 0, and we only ever go towards 0 (marking things not read or write only), never the reverse.

Reply:
I would add an assert in the else case though to confirm this assumption (that MarkedNonReadWriteOnly.contains(VI) is false).